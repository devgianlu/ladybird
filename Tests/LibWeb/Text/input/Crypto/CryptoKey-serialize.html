<!DOCTYPE html>
<script src="../include.js"></script>
<script>
  async function prepareDatabase() {
    const req = window.indexedDB.open("TestDatabase");

    return new Promise((resolve, reject) => {
      req.onerror = () => {
        reject(req.error);
      };
      req.onupgradeneeded = () => {
        req.result.createObjectStore("TestObjectStore");
      };
      req.onsuccess = () => {
        resolve(req.result);
      };
    })
  }

  async function awaitRequest(request) {
    return new Promise((resolve, reject) => {
      request.onerror = () => {
        reject(request.error);
      };
      request.onsuccess = () => {
        resolve(request.result);
      };
    });
  }

  asyncTest(async (done) => {
    for (const [algo, usages] of [
      [{name: "AES-GCM", length: 256}, ["encrypt", "decrypt"]],
      [{name: "AES-CBC", length: 256}, ["encrypt", "decrypt"]],
      [{name: "AES-KW", length: 256}, ["wrapKey", "unwrapKey"]],
      [{name: "HMAC", hash: {name: "SHA-256"}}, ["sign", "verify"]],
      [{name: "RSASSA-PKCS1-v1_5", hash: {name: "SHA-256"}}, ["sign", "verify"]],
      [{name: "ECDSA", namedCurve: "P-256"}, ["sign", "verify"]],
      [{name: "Ed25519"}, ["sign", "verify"]],
    ]) {
      const originalKey = await crypto.subtle.generateKey(algo, true, usages);

      const db = await prepareDatabase()

      const tx = db.transaction("TestObjectStore", "readwrite");
      const store = tx.objectStore("TestObjectStore");
      await awaitRequest(store.put(originalKey, "testKey"));

      const serializedKey = await awaitRequest(store.get("testKey"));

      const originalKeyJwk = await crypto.subtle.exportKey("jwk", originalKey);

      try {
        const serializedKeyJwk = await crypto.subtle.exportKey("jwk", serializedKey);

        for (const key of Object.keys(originalKeyJwk)) {
          println(`${algo.name} ${key}: ${originalKeyJwk[key] === serializedKeyJwk[key]}`);
        }
      } catch (e) {
        println(`${algo.name}: ${e}`);
      }
    }

    done();
  });
</script>
