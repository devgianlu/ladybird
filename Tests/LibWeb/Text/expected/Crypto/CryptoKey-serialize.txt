AES-GCM kty: true
AES-GCM key_ops: false
AES-GCM alg: false
AES-GCM ext: true
AES-GCM k: false
Caught error while running async test: OperationError: unclear key size
